ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+            auto smplr = std::get<1>(bm);
+
+            std::tie(node->texture.image, node->texture.sampler, node->texture.scale) = std::make_tuple(createDerivMap(img.get(), smplr.get()), std::move(smplr), std::get<2>(bm));
+            bsdfQ.push(current->bumpmap.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::COATING:
+        case CElementBSDF::ROUGHCOATING:
+        {
+            nextSym = ir->allocNode<IR::CCoatingBSDFNode>();
+            nextSym->children.count = 1u;
+
+            const float eta = current->dielectric.intIOR/current->dielectric.extIOR;
+
+            auto* node = static_cast<IR::CCoatingBSDFNode*>(nextSym);
+            node->thickness = current->coating.thickness;
+            getSpectrumOrTexture(current->coating.sigmaA, node->sigmaA);
+            node->eta = IR::INode::color_t(eta);
+            node->shadowing = IR::CCoatingBSDFNode::EST_SMITH;
+            if (currType == CElementBSDF::ROUGHCOATING)
+            {
+                node->ndf = ndfMap[current->coating.distribution];
+                getFloatOrTexture(current->coating.alphaU, node->alpha_u);
+                if (node->ndf == IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY)
+                    getFloatOrTexture(current->coating.alphaV, node->alpha_v);
+                else
+                    node->alpha_v = node->alpha_u;
+            }
+            else
+            {
+                node->setSmooth();
+            }
+            bsdfQ.push(current->coating.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::BLEND_BSDF:
+        {
+            nextSym = ir->allocNode<IR::CBSDFBlendNode>();
+            nextSym->children.count = 2u;
+
+            auto* node = static_cast<IR::CBSDFBlendNode*>(nextSym);
+            getFloatOrTexture(current->blendbsdf.weight, node->weight);
+
+            bsdfQ.push(current->blendbsdf.bsdf[1]);
+            bsdfQ.push(current->blendbsdf.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::MIXTURE_BSDF:
+        {
+            nextSym = ir->allocNode<IR::CBSDFMixNode>();
+            auto* node = static_cast<IR::CBSDFMixNode*>(nextSym);
+            const size_t cnt = current->mixturebsdf.childCount;
+            nextSym->children.count = cnt;
+            for (int32_t i = cnt-1u; i >= 0; --i)
+                node->weights[i] = current->mixturebsdf.weights[i];
+
+            for (int32_t i = cnt-1u; i >= 0; --i)
+                bsdfQ.push(current->mixturebsdf.bsdf[i]);
+        }
+        break;
+        }
+
+        IR::INode* newParent = nullptr;
+        if (nextSym->children)
+            nodeQ.push(nextSym);
+        if (!--childrenCountdown && !nodeQ.empty())
+        {
+            newParent = nodeQ.front();
+            nodeQ.pop();
+        }
+
+        parent->children[childrenCountdown] = nextSym;
+        if (newParent)
+        {
+            parent = newParent;
+            childrenCountdown = parent->children.count;
+        }
```

Comment:
why is the logic here so convoluted?

Reply:
Unfortunately i have to agree, it is. Because i couldnt get it done in a prettier way

Reply:
ok once all the issues are dealt with, we'll revisit